### PR TITLE
[FIX] website: prevents the deletion of the form's default values

### PR DESCRIPTION
**Pull request by Guillaume-gdi**
Some form fields are dynamically filled. As the inputs that hold these
fields do not always have a value property, their HTML value attribute
can be modified. Therefore the form option deletes the values so that
they do not become default values (see [1]). Unfortunately this system
is problematic and makes it possible to delete a default value. This
PR ensures that only auto-fillable fields and values from data-for
values will be removed by the option afterwards.

This PR also gives the priority to the auto-fill over the default
value.
Tests are added with this PR to verify that
- when a form field has both a default value and auto-fill, the
  auto-fill has the priority.
- the default values are not erased on later form edit.
- even if the application form is updated, the job_id does not become a
  default value.

Step to reproduce the fixed issue:
- Create a form, add a (required hidden) field with a default value
- Save, then enter edit mode again
- Add a new field and save again

The default values of the form are gone. Enter edit mode again to see
the hidden input, you will see there is no value.

While fixing the bug, another bug has been discovered. Fields that can be auto-filled are no longer auto-filled if certain options are applied to them. This PR contains a commit that fixes this problem as well and improves a test to detect if this problem occurs again.
Steps to reproduce this bug:
 - Go to the "contact us" page or drop a form in a page
 - Edit an auto-fillable field of the form ("Your Name" for example)
 - Change the label position of the field
 - Save

The field is no longer auto-fillable.

[1]: https://github.com/odoo/odoo/commit/b637a5e32f767b62736241042f88fa0cecf9f10b

task-2715201
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by rdeodoo** [comment]
Did not checked the code, but can you add the step to reproduce to your commit?
Should always been one, makes things way easier.. Especially in this case where I wrote a clean one on the task :)
Functionally, the fix is working fine.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Commit message:
- Title: must complete the sentence "When applied, this commit will..." -> so no "s" at the end of "prevents"
- Markdown reference: `[x]:` with a `:` and it should be a full link to a *commit* (ideally not a PR otherwise you don't find the commit that was actually merged, etc etc).

> It is considered that the default value should have priority over the auto-fill

In fact, I think the auto-fill should take priority: if we know the value for the current visitor there is no point to use a default value that was configured for anonymous visitors.

**Comment by qsm-odoo** [comment]
@Guillaume-gdi Also, adding / completing a test could be nice here

File: addons/website/static/src/snippets/s_website_form/000.js [hunk 1]
```diff
@@ -134,10 +134,11 @@ odoo.define('website.s_website_form', function (require) {
                     var $field = self.$target.find('input[name="' + field + '"], textarea[name="' + field + '"]');
                     if (!$field.val() && _.has(dataForValues, field) && dataForValues[field]) {
                         $field.val(dataForValues[field]);
+                        $field.data('website_form_original_default_value', $field.val());
                     } else if (!$field.val() && $field.data('fill-with')) {
                         $field.val(self.preFillValues[$field.data('fill-with')] || '');
+                        $field.data('website_form_original_default_value', $field.val());
```
[the remark this answers is not in the document]
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_website_form/000.js]: The piece of code you point out is useful, it was added [here](https://github.com/odoo/odoo/pull/50745/commits/48475eb6d879c885baa14bfee3130add019d6418) so that when you edit a form to apply for a job, you don't save the job_id as a default value. This job_id must be different for each job. So we have to delete its value => maybe it can be done differently but we can't just delete this piece of code.

This piece of code deletes all the values of the fields which have the property `website_form_original_default_value`. My solution is to put this property only when it is relevant.

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_website_form/000.js]: Neither the explanation nor the fix you mention seems good. As as explained above, the code [here](https://github.com/odoo/odoo/pull/50745/commits/48475eb6d879c885baa14bfee3130add019d6418#diff-e712b94d96fb636dd30632e9540fa8caa06ac145c2618544007b5f29ba77b364R66) has no impact on the DOM. So [doing that](https://github.com/odoo/odoo/pull/50745/commits/48475eb6d879c885baa14bfee3130add019d6418#diff-ec70bde754e307ff6e9c3f50d61e703606fc40619e066464c99ffaddfdbb90aeR442) has simply no effect (again, as I explained above). So if NLE made this code I guess it solves something but I'll need more explanation to understand why. And I really doubt this was the right fix, even by chance. Can you investigate and explain to me why what I explained above is wrong?

[the remark this answers is not in the document]
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_website_form/000.js]: Ok I investigated and here is the point: when you do `.val('XXX')` on a `input` with `type="hidden"` then the value attribute will be set to _XXX_. (That's the case for the job_id stuff)
You can check it [here](https://jsfiddle.net/gdi_odoo/f1j7apk8/1/)

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_website_form/000.js]: Thanks, indeed :+1: That was missing from commit message and code comments for the NLE's script: it was really useful to investiguate. Now can't this be changed ?

A) At best, the public widget (000.js) should not modify the DOM on start (only the value properties like I thought it was doing). For the hidden fields, maybe the solution would be to auto-fill them on submit rather than on page load (that would thus not impact the edition at all).

B) If not possible for some reason (why?), I think NLE's fix has to be reviewed to:
    1. Add more comments explaining why following your investigation
    2. It's always bad to have code in public widgets that is cleaned by the editor. Normally I think you could have done what the cleanForSave does in the public widget's destroy method (thus keeping the code together and probably simplifying it (you can just memorize the inputs there is to clean on start and loop through them on destroy)

File: addons/website/static/src/snippets/s_website_form/000.js [hunk 2]
```diff
@@ -126,18 +126,26 @@ odoo.define('website.s_website_form', function (require) {
...
-                    if (!$field.val() && _.has(dataForValues, field) && dataForValues[field]) {
-                        $field.val(dataForValues[field]);
-                    } else if (!$field.val() && $field.data('fill-with')) {
+                    const inputType = $field.prop('type');
+                    if ($field.prop('tagName') === 'INPUT' &&
+                    inputsWithoutValueProperty.includes(inputType)) {
+                        // In order to keep the DOM intact, we can't fill value
+                        // attribute, only value property.
+                        return;
+                    }
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_website_form/000.js:143]: - Is there a real need to check all those types? Is not the problematic one only "hidden"? Do we currently allow prefilling on checkboxes or worse: submit buttons?
- The comment is not bad but is not explicit enough to me:
    - "keep the DOM intact", replace "intact" by "untouched"
    - "we can't fill value attribute", well you can, you *should* not... but why? It needs explanations.
    - Someone reading this might say "omg it's bad, we cannot prefill hidden fields"... you should explain that it is done on send
    => Something like "Avoid altering the DOM for the pre-filling of input values (interact only with input value *property*) so that those are not accidentally saved during edition. The pre-filling of hidden inputs (which forces a DOM alteration) will only be done on form submit." ? 

Actually, reading that version, I wonder if we can do that as a stable fix: a form with a hidden input as no more value during page life if it was prefilled before... custo may rely on this. It may actually even be bad for master. When I suggested to do this, I had in mind hidden inputs had a value property too, I did not expect the solution would lead to have hidden input with undefined values up until they are send. Your original solution (with many more comments) is probably better then (just setting `website_form_original_default_value` when prefilled via data-for or prefill rpc)). Actually the code which removes those values https://github.com/odoo/odoo/commit/b637a5e32f767b62736241042f88fa0cecf9f10b#diff-ec70bde754e307ff6e9c3f50d61e703606fc40619e066464c99ffaddfdbb90aeR439 -> I am still not sure why it is in `cleanForSave` rather than in the publicWidget destroy though.

So: if you agree, restore what you did before but explain all this with inline code comments :+1: Sorry for the back and forth, that's the R in R&D I guess :grimacing:

File: addons/website/static/tests/tours/website_form_editor.js [hunk 3]
```diff
@@ -314,6 +323,21 @@ odoo.define('website.tour.form_editor', function (require) {
...
+                const inputEl = document.querySelector('.s_website_form_field input');
+                if (inputEl.getAttribute('value') !== 'John Smith') {
+                    console.error('The value attribute is not correctly set');
+                }
+                setTimeout(function () { // Wait for the auto-fillable fields to be filled.
+                    if (inputEl.value !== 'Mitchell Admin') {
+                        console.error('The value property is not correctly set');
+                    }
+                }, 1000);
+            },
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/website_form_editor.js:340]: This step is the last of your tour. If you add a setTimeout inside... well it will never be considered as the tour will be marked as finished before reaching the setTimeout code (or if not, it's really by chance). Did you actually try to break the prefill system to see if your tour was indeed red in that case? 

Also: `const inputEl = document.querySelector('.s_website_form_field input');`is not it the `trigger` you declared? You don't need to research it again, you can use `this.$anchor` or something like that.

Anyways, this whole run can actually be replaced by a better trigger which do not need a setTimeout (which would slow down runbot builds):
`trigger: '.s_website_form_field:eq(0) input[value="John Smith"]:propValue("Mitchell Admin")'`
-> That way that step will be marked as ok as soon as an input with the John Smith default value but the Mitchell Admin actual value is found. (propValue is something of ours). Double check what I say but it should work. I have a slight worry that the prefill done in JS might not trigger a step check in the tour system but that would need to be fixed in that case.

File: addons/website/static/tests/tours/website_form_editor.js [hunk 4]
```diff
@@ -349,7 +373,7 @@ odoo.define('website.tour.form_editor', function (require) {
         {
             content:  "Check if required fields were detected and complete the Subject field",
             extra_trigger:  "form:has(#s_website_form_result.text-danger)" +
-                            ":has(.s_website_form_field:has(label:contains('Your Name')).o_has_error)" +
+                            ":has(.s_website_form_field:has(label:contains('Your Name')):not(.o_has_error))" +
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/website_form_editor.js:376]: Mmh, why did this change?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/tests/tours/website_form_editor.js (hunk 4 above)
Comment (Guillaume-gdi) [reply addons/website/static/tests/tours/website_form_editor.js:365]: Just because I needed to test the "value" attribute/property stuff. So I put a default value on the "Your Name" field (which is auto-fill). Since it has a value, when you submit it is not marked as "error" anymore.  If you prefer I can add a step that empties the content of the field (the default value).

Comment (qsm-odoo) [reply addons/website/static/tests/tours/website_form_editor.js:365]: Oh. Ok. Well at least the previous label saying ""Try to send empty form" should have been adapted then, that's what confused me :grimacing: I guess whatever is fine here, as long as you either change that label or add a step to empty the form.

File: addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js [hunk 5]
```diff
@@ -3,38 +3,114 @@ odoo.define('website_hr_recruitment.tour', function(require) {
 
     var tour = require("web_tour.tour");
 
-    tour.register('website_hr_recruitment_tour', {
```
Comment (qsm-odoo) [inline addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js:6]: Not entirely convinced we should rename this tour in stable. Technically, you could rely on it. But I am not entirely sure to understand your double tour with double application. You made an utility method to apply for a job, why not calling it twice in a unique tour with the same name?  Launching two tours is way slower, you need to spawn a chrome headless process etc. The only times we call `start_tour` multiple times is when there a need for some python code between the two calls or that you have to switch the user which runs the tour.

File: addons/website/static/tests/tours/website_form_editor.js [hunk 6]
```diff
@@ -314,14 +323,18 @@ odoo.define('website.tour.form_editor', function (require) {
         {
             content:  "Wait reloading...",
             trigger:  "html:not(:has(#completlyloaded)) div",
+        },
+        {
+            content: 'Verify value attribute and property',
+            trigger: '.s_website_form_field:eq(0) input[value="John Smith"]:propValue("Mitchell Admin")',
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/website_form_editor.js:329]: It's cool to have this test about default value but does this cover the bug you are fixing in this PR? From what I understand the bug occurs on re-edition of the form which already has a default value, this test does not test that, right? :thinking:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/tests/tours/website_form_editor.js (hunk 6 above)
Comment (qsm-odoo) [reply addons/website/static/tests/tours/website_form_editor.js:329]: Well in fact the commit should indeed be split, I was going to accept that it's not but it really is confusing. So yes split the fix and the auto-fill priority in two separate commits. Actually not sure if it should be for stable or not, but in this case, I would keep it for stable. So split in tho commits, in this PR.

Still, I think the test should be further improved to test the bug you fix here (by reediting the form that has a default value).

File: addons/website/static/src/snippets/s_website_form/000.js [hunk 7]
```diff
@@ -130,14 +130,23 @@ odoo.define('website.s_website_form', function (require) {
             if ($dataFor.length || !_.isEmpty(this.preFillValues)) {
                 var dataForValues = $dataFor.length ? JSON.parse($dataFor.data('values').replace('False', '""').replace('None', '""').replace(/'/g, '"')) : [];
                 var fields = _.pluck(this.$target.serializeArray(), 'name');
+                // All types of inputs do not have a value property (eg:hidden),
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_website_form/000.js:133]: Commit message: Why do you not add the full github link as a reference..?
image

That's the main point of making references instead of adding the link directly inside the text: being able to see the full link outside of github (git log) and on a dedicated line since the links are long. So in general, replace that by the full link.

However, here it is worse, you link does not actually redirect to anything (maybe since you use a small version of the hash it now conflicts? Even more reason to put the full link): 
image

File: addons/website/static/src/snippets/s_website_form/000.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_website_form/000.js]: > 2. It's always bad to have code in public widgets that is cleaned by the editor. Normally I think you could have done what the cleanForSave does in the public widget's destroy method (thus keeping the code together and probably simplifying it (you can just memorize the inputs there is to clean on start and loop through them on destroy)

@Guillaume-gdi What about this? I can't remember you proved me this was wrong? (not needing `cleanForSave` like you explain in your code comment above)

File: addons/website/static/tests/tours/website_form_editor.js [hunk 8]
```diff
@@ -314,14 +323,40 @@ odoo.define('website.tour.form_editor', function (require) {
...
+        {
+            content: 'Enter in edit mode again',
+            trigger: 'a[data-action="edit"]',
+            run: 'click',
+        },
+        {
+            content: 'Verify that the editor appeared',
+            trigger: 'button[data-action="save"]',
+            run: function () {}
+        },
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/website_form_editor.js:341]: This does not need to be an extra step, at worst you should add it as an extra_trigger of the next step

File: addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js [hunk 9]
```diff
@@ -2,39 +2,114 @@ odoo.define('website_hr_recruitment.tour', function(require) {
     'use strict';
 
     var tour = require("web_tour.tour");
+    function applyForAJob(jobName, applicantInformation) {
```
Comment (qsm-odoo) [inline addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js:5]: It makes sense to have put this in a dedicated commit but then add more reference in the commit message (mentioning the commit that actually fixed the bug for which you add a test otherwise, we wonder why you add new test in stable without fixing anything).

File: addons/website/static/tests/tours/website_form_editor.js [hunk 10]
```diff
@@ -314,14 +323,40 @@ odoo.define('website.tour.form_editor', function (require) {
...
+        ...addCustomField('many2one', 'select', 'Select Field', true),
+        {
+            content: 'Save the page',
+            trigger: 'button[data-action=save]',
+            run: 'click',
+        },
+        {
+            content: 'Verify that the value has not been deleted',
+            trigger: '.s_website_form_field:eq(0) input[value="John Smith"]',
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/website_form_editor.js:355]: In the last commit I see this:
image
Is this actually needed to wait for the reload this way (and if so, should not it be the same here?)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_website_form/000.js [hunk 11]
```diff
@@ -141,10 +141,10 @@ odoo.define('website.s_website_form', function (require) {
                 // customizations that use the current behavior as a feature.
                 for (const name of fieldNames) {
                     const fieldEl = this.$target[0].querySelector(`[name="${name}"]`);
-                    if (!fieldEl.value && dataForValues && dataForValues[name]) {
+                    if (dataForValues && dataForValues[name]) {
                         fieldEl.value = dataForValues[name];
                         this.filledFields.push(fieldEl);
-                    } else if (!fieldEl.value && this.preFillValues[fieldEl.dataset.fillWith]) {
+                    } else if (this.preFillValues[fieldEl.dataset.fillWith]) {
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_website_form/000.js:147]: @Guillaume-gdi In the first commit, you add in `destroy` a reset of the value for the `filledFields`. You reset them via `.value = ''`. In this third commit, you allow prefill values to replace existing default values (not empty ones)... but on destroy you still do `.value = ''`. It works most of the time as `.value = ''` only writes on the property... but in the case of an hidden input with a configured default value, you probably still have the bug of losing the configured default value if it was prefilled first, is not it? :grimacing: 

Probably worth an extra check in the website_hr_recruitment test in this commit?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_website_form/000.js [hunk 12]
```diff
@@ -126,19 +124,31 @@ odoo.define('website.s_website_form', function (require) {
...
+                // this, we have to clean up these values at destroy or else the
+                // data loaded here could become default values. We could set
+                // the values to submit() for these fields but this could break
+                // customizations that use the current behavior as a feature.
+                for (const name of fieldNames) {
+                    const fieldEl = this.$target[0].querySelector(`[name="${name}"]`);
+                    this.initialValues.push({
+                        fieldEl,
+                        initialValue: fieldEl.getAttribute('value') || '',
+                    });
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_website_form/000.js:145]: That's fine like this, but FYI I would have use a Map.

File: addons/website/static/src/snippets/s_website_form/000.js [hunk 13]
```diff
@@ -194,6 +204,9 @@ odoo.define('website.s_website_form', function (require) {
             // All 'hidden if' fields start with d-none
             this.$target[0].querySelectorAll('.s_website_form_field_hidden_if:not(.d-none)').forEach(el => el.classList.add('d-none'));
 
+            // Reset the initial default values.
+            this.initialValues.forEach(el => el.fieldEl.setAttribute('value', el.initialValue));
+
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_website_form/000.js:209]: I was already scared that you register all initials values (even the ones that are not prefilled) but if you do it this way, I am pretty sure this will leave `value=""` on every field at reedition... :grimacing: Or am I wrong?

File: addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js [hunk 14]
```diff
@@ -2,39 +2,131 @@ odoo.define('website_hr_recruitment.tour', function(require) {
...
+    }, {
+        content: 'Enter in edit mode',
+        trigger: 'a[data-action="edit"]',
+    }, {
+        content: 'Add a fake default value for the job_id field',
+        trigger: 'button[data-action="save"]',
+        run: function () {
+            // It must be done in this way because the editor does not allow to
+            // put a default value on a hidden field.
+            document.querySelector('input[name="job_id"]').value = 'FAKE_JOB_ID_DEFAULT_VAL';
```
Comment (qsm-odoo) [inline addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js:84]: I hope what you say is not true otherwise our hidden field feature makes no sense... why would we allow to add hidden fields if we cannot put values in them?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js (hunk 14 above)
Comment (Guillaume-gdi) [reply addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js:84]: Mmhh I think my comment is indeed confusing. The field I'm talking about here is a field with a `type=hidden`, which is different from the visibility feature of the field = hidden which leaves the field type unchanged but puts a class `s_website_form_field_hidden`. The `type=hidden` fields remain hidden in edit mode via the editor so impossible to set a default value to them.

File: addons/website/static/src/snippets/s_website_form/000.js [hunk 15]
```diff
@@ -126,19 +124,36 @@ odoo.define('website.s_website_form', function (require) {
...
+                        newValue = this.preFillValues[fieldEl.dataset.fillWith];
                     }
-                    $field.data('website_form_original_default_value', $field.val());
-                });
+                    if (newValue) {
+                        this.initialValues.push({
+                            fieldEl,
+                            initialValue: fieldEl.getAttribute('value') || '',
+                            hasValueAttr: fieldEl.hasAttribute('value'),
+                        });
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_website_form/000.js:153]: Still not good enough: you now declare redundant information. Why do you need `initialValue` and `hasValueAttr`? Actually, if `initialValue` is falsy, then it's always better to remove the value attribute no? So it seems to be more code for a lesser behavior.

Also, use a `Map` as I suggested to map the field to its initial value. It makes way more sense than an array of objects.

File: addons/website/static/src/snippets/s_website_form/000.js [hunk 16]
```diff
@@ -126,19 +124,36 @@ odoo.define('website.s_website_form', function (require) {
...
-                _.each(fields, function (field) {
-                    var $field = self.$target.find('input[name="' + field + '"], textarea[name="' + field + '"]');
-                    if (!$field.val() && _.has(dataForValues, field) && dataForValues[field]) {
-                        $field.val(dataForValues[field]);
-                    } else if (!$field.val() && $field.data('fill-with')) {
-                        $field.val(self.preFillValues[$field.data('fill-with')] || '');
+            const dataForEl = document.querySelector(`[data-for='${this.$target[0].id}']`);
+            if (dataForEl || Object.keys(this.preFillValues).length) {
+                const dataForValues = dataForEl ? JSON.parse(dataForEl.dataset.values
+                        .replace('False', '""').replace('None', '""').replace(/'/g, '"')) : {};
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_website_form/000.js:130]: Seems strangely indented actually

File: addons/website/static/src/snippets/s_website_form/000.js [hunk 17]
```diff
@@ -194,6 +209,14 @@ odoo.define('website.s_website_form', function (require) {
             // All 'hidden if' fields start with d-none
             this.$target[0].querySelectorAll('.s_website_form_field_hidden_if:not(.d-none)').forEach(el => el.classList.add('d-none'));
 
+            // Reset the initial default values.
+            for (const el of this.initialValues) {
+                el.fieldEl.setAttribute('value', el.initialValue);
+                if (!el.hasValueAttr) {
+                    el.fieldEl.removeAttribute('value');
+                }
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_website_form/000.js:217]: See even here, it leads to strange code: you set a value then potentially remove it. While
```javascript
if (value) {
   // setValue
} else {
   // removeValue
}
```
Is really explicit and better.

File: addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js [hunk 18]
```diff
@@ -2,39 +2,131 @@ odoo.define('website_hr_recruitment.tour', function(require) {
...
+            }
+    },
+    applyForAJob('Internship',
+        {
+            name: 'Jack Doe',
+            email: 'jack@doe.com',
+            phone: '118.712',
+            subject: '### HR [INTERN] RECRUITMENT TEST DATA ###',
+        }),
+    ));
```
Comment (qsm-odoo) [inline addons/website_hr_recruitment/static/tests/tours/website_hr_recruitment.js:61]: Hard to read in review but your use of concat seems off or maybe I am confused by the indentation. Aren't you giving an object as second argument? How can that work?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@Guillaume-gdi Ping me once your done later today, no need to move this back to RD Dev, thanks

File: addons/website/static/src/snippets/s_website_form/000.js (hunk 15 above)
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_website_form/000.js:153]: In fact I wanted to avoid that inputs like this `<input value="" />` or like this `<input value />` change but indeed it's probably not an issue if those inputs lost their value attribute

Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_website_form/000.js:153]: It's more than not an issue, it's better :wink:

**Comment by Guillaume-gdi** [comment]
@qsm-odoo I applied the changes you asked me to make :)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@Guillaume-gdi Something's still not right :/ I was about to merge but then went and test your tours. The second commit explains that it adds a test that an old commit should have added, good. So my first logical try is:
- Go to 15.0
- Only add your second commit
- Launch the test to see that it passes and maybe after that, revert that old commit and see that your test fails.

Well your test does not pass... because it depends on your first commit I guess. So I told myself "ok that should have been explained in the commit message but that's ok". So I went and did the same for 15.0 + your first commit + your second commit... and there your test does not pass again... probably because it depends on your third commit. This is not right. I should be able to merge your PR commit by commit if I wanted too. If another team were ever to revert one our your commit for some reason... they would have to know they also have to revert part of some other commits so that it works.

This needs refactoring again, what I want:
- Put your current second commit as the first commit: you add a test to complete an old commit, with the necessary steps
- Put your current first commit as the second one: you can then complete the recruitment test with the added/modified steps if it makes sense
- Same for third commit: you can test complete the recruitment test with the added/modified steps if it makes sense
- Same for fourth commit

After each commit, I want the tests that are modified by this commit to fail without the rest of the commit and succeed with the full commit. Except for the first one of course, that I want to fail without the old commit you mention in your messages. In the end, I want that the whole PR diff stays the same but that each commit makes sense on its own.

Also, I can't remember if we already discussed this but that old commit you mentioned was merged in 14.0, should not (some of?) your commits be made over there? Thank you.

**Comment by Guillaume-gdi** [comment]
> Also, I can't remember if we already discussed this but that old commit you mentioned was merged in 14.0, should not (some of?) your commits be made over there?
By double checking, I saw that the commit I mentioned was not the original. The original was on 13.0. Indeed it makes sense that the test that should have been added with the mentioned fix is in a PR that targets the same version so here it is: https://github.com/odoo/odoo/pull/89220 
For the other commits, merging them here (in 15.0) seems coherent to me since the default values and auto-fill appeared in 15.0.
Once [this one](https://github.com/odoo/odoo/pull/89220) merged, I'll do a pull rebase on this PR and I'll remove the first commit from this PR.

**Comment by Guillaume-gdi** [comment]
@qsm-odoo now this one should be OK :+1:

**Review by qsm-odoo — APPROVED** [review]
I made this final change otherwise it could crash on destroy: https://github.com/odoo/odoo/compare/19db02d74ebdc7950f997c32341c3643e3dc7357..4f62844e54f3d5eec1b5f74c5d68837bbee52a5b

@robodoo rebase-ff r+
